File: 프로그래머스/3/43238. 입국심사/입국심사.py
def solution(n, times):
    left = 1
    right = n * max(times)
    answer = right
    
    while left <= right:
        mid = (left + right) // 2
        
        total_sum = sum([mid // time for time in times])
        
        if total_sum >= n:
            # 시간 더 줄여도 됨
            answer = mid
            right = mid - 1
        else:
            left = mid + 1
        
    return answer

File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.*;

class Solution {
    public long solution(int n, int[] times) {
        long left = 1;
        long right = (long) n * Arrays.stream(times).max().getAsInt();
        long answer = right;
        
        while(left<=right){
            long mid = (left + right) / 2; // 끝낼 시간
            long count = 0;
            for(int time : times){
                count += mid / (long) time;
            }
            if(count >= n){ // 시간 더 줄여봐도 됨
                right = mid - 1;
                answer = mid;
            }else{
                left = mid + 1;
            }
        }
        
        return answer;
    }
}